website/src/plugins/changelog/index.js

```diff
@@ -0,0 +1,149 @@
+ */
+/* eslint-disable import/no-extraneous-dependencies */
+
+const path = require('path');
+const fs = require('fs-extra');
+const pluginContentBlog = require('@docusaurus/plugin-content-blog');
+const {aliasedSitePath, docuHash} = require('@docusaurus/utils');
+const syncAvatars = require('./syncAvatars');
+
+/**
+ * Multiple versions may be published on the same day, causing the order to be
+ * the reverse. Therefore, our publish time has a "fake hour" to order them.
+ */
+const publishTimes = new Set();
+/** @type {Record<string, {name: string, url: string, alias: string, imageURL: string}>} */
+const authorsMap = {};
+
+/**
+ * @param {string} section
+ */
+function processSection(section) {
+  const title = section
+    .match(/\n## .*/)?.[0]
+    .trim()
+    .replace('## ', '');
+  if (!title) {
+    return null;
+  }
+  const content = section
+    .replace(/\n## .*/, '')
+    .trim()
+    .replace('running_woman', 'running');
+
+  let authors = content.match(/## Committers: \d+.*/ms);
+  if (authors) {
+    authors = authors[0]
+      .match(/- .*/g)
+      .map(
+        (line) =>
+          line.match(
+            /- (?:(?<name>.*?) \()?\[@(?<alias>.*)\]\((?<url>.*?)\)\)?/,
+          ).groups,
+      )
+      .map((author) => ({
+        ...author,
+        name: author.name ?? author.alias,
+        imageURL: `./img/${author.alias}.png`,
+      }))
+      .sort((a, b) => a.url.localeCompare(b.url));
+
+    authors.forEach((author) => {
+      authorsMap[author.alias] = author;
+    });
+  }
+  let hour = 20;
+  const date = title.match(/ \((.*)\)/)[1];
+  while (publishTimes.has(`${date}T${hour}:00`)) {
+    hour -= 1;
+  }
+  publishTimes.add(`${date}T${hour}:00`);
+
+  return {
+    title: title.replace(/ \(.*\)/, ''),
+    content: `---
+date: ${`${date}T${hour}:00`}${
+      authors
+        ? `
+authors:
+${authors.map((author) => `  - '${author.alias}'`).join('\n')}`
+        : ''
+    }
+---
+
+# ${title.replace(/ \(.*\)/, '')}
+
+<!-- truncate -->
+
+${content.replace(/####/g, '##')}`,
+  };
+}
+
+/**
+ * @param {import('@docusaurus/types').LoadContext} context
+ * @returns {import('@docusaurus/types').Plugin}
+ */
+async function ChangelogPlugin(context, options) {
+  const generateDir = path.join(context.siteDir, 'changelog/source');
+  const blogPlugin = await pluginContentBlog.default(context, {
+    ...options,
+    path: generateDir,
+    id: 'changelog',
+    blogListComponent: '@theme/ChangelogList',
+    blogPostComponent: '@theme/ChangelogPage',
+  });
+  const changelogPath = path.join(__dirname, '../../../../CHANGELOG.md');
+  return {
+    ...blogPlugin,
+    name: 'changelog-plugin',
+    async loadContent() {
+      const fileContent = await fs.readFile(changelogPath, 'utf-8');
+      const sections = fileContent
+        .split(/(?=\n## )/ms)
+        .map(processSection)
+        .filter(Boolean);
+      await Promise.all(
+        sections.map((section) =>
+          fs.outputFile(
```

Comment:
that works, but wonder if we really want to encourage users to write files in this lifecycle in the long term, particularly for CMS integrations.

I'd like to explore something like this: https://github.com/sysgears/webpack-virtual-modules

Reply:
Mmm, we not only read these files during webpack build, but during server-side processing as well (e.g. to extract description, authors...). I think this would make life easier for everyone.